compiler/rustc_mir/src/borrow_check/diagnostics/conflict_errors.rs

```diff
@@ -1268,6 +1268,24 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
 
         if return_span != borrow_span {
             err.span_label(borrow_span, note);
+
+            let tcx = self.infcx.tcx;
+
+            let suggestion = match tcx.sess.source_map().span_to_snippet(return_span) {
+                Ok(string) => string.contains(".iter()") && !string.contains(".collect"),
+                Err(_) => false,
+            };
+
+            // If the snippet contains ".iter()", then
+            // make `.collect()` suggestion.
```

Comment:
This seems like the wrong approach; it's dependent on formatting and also breaks if the iterator was returned from a function. Try checking if the borrowed value implements `Iterator` instead - maybe you can use `borrow.borrowed_place.ty()`?

Reply:
The function `type_implements_trait` sounds promising (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.type_implements_trait)

But I am struggling to find a way to get the `DefId` of the `Iterator` trait. Do you know how I do that?

Reply:
Hmm, I think you could add `rustc_diagnostic_item` to `Iterator`, the same way it's there for `IntoIterator`. Then you can get it from https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.diagnostic_items.